mypy/build.py

```diff
@@ -715,46 +721,125 @@ def stats_summary(self) -> Mapping[str, object]:
+        manager.log("Error writing fine-grained extra deps JSON file {}".format(cache_path))
+        error = True
+    else:
+        fg_deps_meta['@extra'] = {'path': cache_path, 'mtime': manager.getmtime(cache_path)}
+
+    for id in rdeps:
+        _, _, deps_json = get_cache_names(id, graph[id].xpath, manager)
+        assert deps_json
+        manager.log("Writing deps cache", deps_json)
+        if not manager.metastore.write(deps_json, deps_to_json(rdeps[id])):
+            manager.log("Error writing fine-grained deps JSON file {}".format(deps_json))
+            error = True
+        else:
+            fg_deps_meta[id] = {'path': deps_json, 'mtime': manager.getmtime(deps_json)}
+
+    meta_snapshot = {}  # type: Dict[str, str]
     for id, st in graph.items():
         # If we didn't parse a file (so it doesn't have a
         # source_hash), then it must be a module with a fresh cache,
         # so use the hash from that.
         if st.source_hash:
-            meta_snapshot[id] = st.source_hash
+            hash = st.source_hash
         else:
             assert st.meta, "Module must be either parsed or cached"
-            meta_snapshot[id] = st.meta.hash
+            hash = st.meta.hash
+        meta_snapshot[id] = hash
 
-    if not metastore.write(proto_meta, json.dumps(meta_snapshot)):
-        manager.log("Error writing protocol meta JSON file {}".format(proto_cache))
-        error = True
-    listed_proto_deps = {k: list(v) for (k, v) in proto_deps.items()}
-    if not metastore.write(proto_cache, json.dumps(listed_proto_deps)):
-        manager.log("Error writing protocol deps JSON file {}".format(proto_cache))
+    meta = {'snapshot': meta_snapshot, 'deps_meta': fg_deps_meta}
+
+    if not metastore.write(meta_path, json.dumps(meta)):
+        manager.log("Error writing fine-grained deps meta JSON file {}".format(meta_path))
         error = True
+
     if error:
         manager.errors.set_file(_cache_dir_prefix(manager), None)
-        manager.errors.report(0, 0, "Error writing protocol dependencies cache",
+        manager.errors.report(0, 0, "Error writing fine-grained dependencies cache",
                               blocker=True)
 
 
+def invert_deps(
+        deps: Dict[str, Set[str]],
+        graph: Graph) -> Tuple[Dict[str, Dict[str, Set[str]]], Dict[str, Set[str]]]:
+    from mypy.server.target import module_prefix, trigger_to_target
+
+    rdeps = {id: {} for id in graph}  # type: Dict[str, Dict[str, Set[str]]]
+    extra_deps = {}  # type: Dict[str, Set[str]]
+    for trigger, targets in deps.items():
+        module = module_prefix(graph, trigger_to_target(trigger))
+        if module and graph[module].tree:
+            mod_rdeps = rdeps.setdefault(module, {})
+            mod_rdeps.setdefault(trigger, set()).update(targets)
+        else:
+            extra_deps.setdefault(trigger, set()).update(targets)
+
+    return (rdeps, extra_deps)
+
+
+def process_deps(proto_deps: Dict[str, Set[str]],
```

Comment:
The name is not very descriptive and docstring doesn't help.

This actually seems to do a few things:
* Load some fine-grained deps from the cache.
* Calculate fine-grained dependencies for files that were processed in the build (if this is an incremental build, this is a subset of all modules).
* Splits/inverts the dependencies per file.

I'd suggest renaming the function (maybe something like `generate_deps_for_cache`?). Also mention all the above things in the docstring (if they are true). Also it's worth mentioning if this is only used in coarse-grained incremental mode.